functions/src/glean.ts

```diff
@@ -0,0 +1,228 @@
+}
+
+/*
+ * studyEnrollment
+ * Glean ping: study-enrollment
+ */
+export async function studyEnrollment(
+  rallyID: string,
+  studyID: string
+): Promise<void> {
+  const releaseGlean = await gleanLock.acquire();
+  initializeGlean();
+
+  rallyMetrics.id.set(rallyID);
+  enrollmentMetrics.studyId.set(studyID);
+
+  await submitPingFlag.acquire();
+  rallyPings.studyEnrollment.submit();
+  await submitPingFlag.waitForUnlock();
+
+  releaseGlean();
+}
+
+/*
+ * studyUnenrollment
+ * Glean ping: study-unenrollment
+ */
+export async function studyUnenrollment(
+  rallyID: string,
+  studyID: string
+): Promise<void> {
+  const releaseGlean = await gleanLock.acquire();
+  initializeGlean();
+
+  rallyMetrics.id.set(rallyID);
+  unenrollmentMetrics.studyId.set(studyID);
+
+  await submitPingFlag.acquire();
+  rallyPings.studyUnenrollment.submit();
+  await submitPingFlag.waitForUnlock();
+
+  releaseGlean();
+}
+
+/*
+ * Helper function for initializing Glean
+ */
+function initializeGlean(): void {
+  Glean.setDebugViewTag(GLEAN_DEBUG_VIEW_TAG);
+  Glean.setLogPings(true);
+
+  // Glean.initialize is a no-op if Glean is already initialized
+  Glean.initialize(GLEAN_RALLY_APP_ID, true, {
+    appDisplayVersion: GLEAN_APP_DISPLAY_VERSION,
+    plugins: [new PingEncryptionPlugin(GLEAN_ENCRYPTION_JWK)],
+    httpClient: new CustomPingUploader(),
+  });
+}
+
+/*
+ * Helper function for setting user metrics
+ * from demographic data (mapping)
+ */
+function setUserMetrics(data: any): void { // eslint-disable-line @typescript-eslint/no-explicit-any
+  if ("age" in data) {
+    userMetrics.age[`band_${data["age"]}`].set(true);
+  }
+
+  if ("gender" in data) {
+    userMetrics.gender[data["gender"]].set(true);
+  }
+
+  if ("hispanicLatinxSpanishOrigin" in data) {
+    const label =
+      data["hispanicLatinxSpanishOrigin"] === "other"
+        ? "other"
+        : "hispanic_latinx_spanish";
+    userMetrics.origin[label].set(true);
+  }
+
+  if ("race" in data) {
+    for (const raceLabel of data["race"]) {
+      const label =
+        raceLabel === "american_indian_or_alaska_native"
+          ? "am_indian_or_alaska_native"
+          : raceLabel;
+      userMetrics.races[label].set(true);
+    }
+  }
+
+  if ("school" in data) {
+    const KEY_FIXUP: any = { // eslint-disable-line @typescript-eslint/no-explicit-any
+      high_school_graduate_or_equivalent: "high_school_grad_or_eq",
+      some_college_but_no_degree_or_in_progress: "college_degree_in_progress",
+    };
+
+    const originalLabel = data["school"];
+    const label =
+      originalLabel in KEY_FIXUP ? KEY_FIXUP[originalLabel] : originalLabel;
+    userMetrics.school[label].set(true);
+  }
+
+  if ("exactIncome" in data) {
+    userMetrics.exactIncome.set(data["exactIncome"]);
+  }
+
+  if ("zipcode" in data) {
+    userMetrics.zipcode.set(data["zipcode"]);
+  }
+}
+
+/**
+ * Custom Ping Uploader for Glean
+ * TODO: replace direct POST request with Google Cloud Task
```

Comment:
Do you want to do this as a followup PR, could you file an issue for it if so? I think that might make sense given the complexity.

Reply:
I filed a Jira ticket to get the cloud task set up on the GCP side too https://mozilla-hub.atlassian.net/browse/DSRE-750

Reply:
https://github.com/mozilla-rally/rally-web-platform/issues/356